### ISSUE TITLE
[react/destructuring-assignment] callback conflicts with props destructuring

### ISSUE THREAD

**Issue by pilarArr** [yannickcr/eslint-plugin-react#1950]
I'm having issues with this rule when I use callbacks. For example:

```javascript
class Example extends Component {

  callbackMethod = () => {
    const { someProp, anotherProp } = this.props;
    // do stuff with the props
    // now I save anotherProp in the database with function
    functionWithCallback( anotherProp, (error, result) => {
      // when this is saved the props are updated after fetching the 
      // new ones from the database the but if I use 
      // the destructured props from the first statement I get the 
      // old props and if I destructure here I get name conflicts 
    } )
  }

  render() {
    const { foo } = this.props;

    return <div onClick={this.callbackMethod}>{foo.bar}</div>;
  }
}
```

This is my problem with this rule right now and I'm not sure how I should tackle this.

**Comment by ljharb** [comment]
You can rename whole destructuring using a colon.

**Comment by pilarArr** [comment]
Yes, I'm doing that now, but it seems like a stretch and I have to keep track if I'm using the props inside a callback or not to check if I'm destructuring it in the correct place.

```javascript
class Example extends Component {

  callbackMethod = () => {
    const { someProp, anotherProp } = this.props;
    // do stuff with the props
    // now I save anotherProp in the database with function
    functionWithCallback( anotherProp, (error, result) => {
      // when this is saved the props are updated after fetching the 
      // new ones from the database
      // I get new props by destructuring again
      const { someProp: newSomeProp, anotherProp: newAnotherProp } = this.props;
      // do stuff with the props
    } )
  }

  render() {
    const { foo } = this.props;

    return <div onClick={this.callbackMethod}>{foo.bar}</div>;
  }
}
```

**Comment by pilarArr** [comment]
I guess I would have to define the callback in another function to avoid any confusion or conflics 

```javascript
class Example extends Component {

    someCallback = (error, result) => {
      // when this is saved the props are updated after fetching the 
      // new ones from the database
      // I get new props by destructuring again
      const { someProp, anotherProp } = this.props;
      // do stuff with the props
    }

  callbackMethod = () => {
    const { someProp, anotherProp } = this.props;
    // do stuff with the props
    // now I save anotherProp in the database with function
    functionWithCallback( anotherProp, this.someCallback);
  }

  render() {
    const { foo } = this.props;

    return <div onClick={this.callbackMethod}>{foo.bar}</div>;
  }
}
```

**Comment by ljharb** [comment]
If there's still something on the latest version that isn't doing what you think it should, please file a new issue.
